### PR TITLE
First pass at programmer docs

### PR DESCRIPTION
**Pull request by andysim**
## Description
This is the first of a handful of planned PRs to flesh out the programming docs.  I'm happy to take feed back on the general direction early in the review cycle, but note that these are neither proofread nor spellchecked until I comment to say so: please don't waste time on small details until then.

## Checklist
- [x] Provide basic overview of C++/Python/Numpy programming philosophy.

## Status
- [x] Ready for review
- [ ] Ready for **squash** merge

### CODE REVIEW

**Comment by JonathonMisiewicz** [comment]
It looks like something went wrong with the commit. I'm seeing a lot of changes to test cases, including ones I know that I made for findif.

**Comment by loriab** [comment]
I think they're in `samples/`, @JonathonMisiewicz, which only gets regenerated from `tests/` upon docs build.

**Comment by andysim** [comment]
Lori's right (as usual); it's the auto-generated samples that changed.  I thought about removing them from this PR to make the review easier but figured that the samples will need to be updated eventually anyway.  The samples can be completely ignored for review purposes, but I'm happy to roll back the changes if anybody would prefer a more minimal PR.

File: doc/sphinxman/source/prog_ways_to_add.rst [hunk 1]
```diff
@@ -35,15 +35,106 @@ Ways to Add Code: Psi4NumPy, Plugins, Full Integration
...
+:p4numpy:`Psi4NumPy <>` repository.
+
+.. _`sec:prog_plugins`:
+
+Avoiding the need to modify Psi4, using plugins
+-----------------------------------------------
+
+In the early days when |PSIfour| was still primarily a C++ code, development
+was very cumbersome due to a lengthy build process.  To expedite development, a
+plugin system was developed.  This plugin machinery allows develpers to access
```
Comment (robertodr) [inline doc/sphinxman/source/prog_ways_to_add.rst]: Typo: "develper" --> "developers"

File: doc/sphinxman/source/prog_ways_to_add.rst [hunk 2]
```diff
@@ -35,15 +35,106 @@ Ways to Add Code: Psi4NumPy, Plugins, Full Integration
...
+the classes defined in the innards of |PSIfour|, with only the small plugin
+code requiring recompilation during development.  The resulting lightweight
+code can be maintained and distributed independently of |PSIfour|, making this
+a good strategy for development, especially in cases where tighter integration
+of the new code with existing |PSIfour| machinery is required than that
+afforded by the Numpy based stratege outlined in the :ref:`sec:prog_psi4numpy`.
```
Comment (robertodr) [inline doc/sphinxman/source/prog_ways_to_add.rst]: Typo: "stratege" --> "strategy"

**Comment by JonathonMisiewicz** [comment]
I'm fine with leaving the sample updates in.

Looks good to me so far.

**Comment by andysim** [comment]
Ok, I think this one's ready for a first pass of reviews; docs are much better done in bite-sized chunks IMHO.  Feedback is welcomed from everybody.  The current version is designed to provide a reasonable overview without being too comprehensive - from here we can discuss where to dive in deeper.  Note that only first few files need review - as soon as you hit the files in `samples` you're in autogen territory and can stop reading.  Thanks for the early comments, @robertodr; I've fixed those typos as well as a few others.

**Comment by andysim** [comment]
The thing to pay most attention to when reviewing is the code of conduct.  This is mostly the vanilla contributor covenant, but with small additions to emphasize that we value inexperienced devs.

File: doc/sphinxman/source/prog_ways_to_add.rst [hunk 3]
```diff
@@ -35,15 +35,106 @@ Ways to Add Code: Psi4NumPy, Plugins, Full Integration
...
+Python has a rich set of tools for obtaining data from external sources,
+writing this tool in the Python layer is a natural choice.  If we know that the
+matrix will always be small enough to fit in memory, we can simply rely on the
+routines present in NumPy to do the heavy lifting and the code is easy to
+implement entirely in the Python layer.  In the case where the matrix operation
+is non-standard and requires some specialized code to handle disk-based
+storage, the decision to write in Python is less clear cut.  It is certainly
+possible to write these out-of-core routines using Numpy primitives, but there
+are a number of tools in |PSIfour| already to perform tasks like these that are
+required, *e.g.*, for cluster.  In this case, a good design would be to write a
```
Comment (loriab) [inline doc/sphinxman/source/prog_ways_to_add.rst]: is there a "coupled" missing here?

File: doc/sphinxman/source/prog_ways_to_add.rst [hunk 4]
```diff
@@ -35,15 +35,106 @@ Ways to Add Code: Psi4NumPy, Plugins, Full Integration
...
+simple piece of code in the C++ layer that performs the matrix operation on a
+given input, using the I/O routines available in |PSIfour| and the parallelism
+afforded by OpenMP, and to make that code available to the front end as
+described in :ref:`sec:prog_tour-exposing`.  The Python layer could then be
+responsible for obtaining the input data and calling this C++ code to do the
+manipulations, allowing each language layer to handle the subset of the work
+that caters to their individual strengths.
```
Comment (loriab) [inline doc/sphinxman/source/prog_ways_to_add.rst]: its individual strength.

File: doc/sphinxman/source/prog_style_c.rst [hunk 5]
```diff
@@ -124,3 +124,11 @@ Note that users can supply MiB, GB, bytes, or whatever; this guideline is for ou
...
                         (required *  8 / (1024 * 1024 * 1024.0)),
 
+Code format
+-----------
+
+The format of the C++ source and headers follows the style prescribed in
+:source:`.clang-format`.  To enforce this style on all files in the current
+directory, simply run ::
+
+    clang-format -i *.h *.cc
```
Comment (loriab) [inline doc/sphinxman/source/prog_style_c.rst]:
```
To retain hand-formatted code (e.g., for equations or data structures), surround by::

    // clang-format off
    leave_me_alone;
    // clang-format on

    format_me_better;

    /* clang-format off */
    leave_me_along_too;
    /* clang-format on */
```

File: doc/sphinxman/CMakeLists.txt [hunk 6]
```diff
@@ -57,10 +57,10 @@ if(PERL_FOUND AND SPHINX_FOUND AND SPHINX_STUFF_FOUND)
     manage_addon.rst numpy.rst build_planning.rst build_faq.rst
```
Comment (robertodr) [inline doc/sphinxman/CMakeLists.txt]: Just a general question: what do you think about having one source file per line (here and in other `CMakeLists.txt`)? It should make it easier to parse diffs.

File: doc/sphinxman/source/prog_ways_to_add.rst [hunk 7]
```diff
@@ -35,15 +35,106 @@ Ways to Add Code: Psi4NumPy, Plugins, Full Integration
...
-integrated fully into Psi4, or stay as a Plugin.
+Since those early days, the code has undergone some important structural
+changes that have greatly simplified the development workflow.  These changes
+were motivated by the realization that only a few bottlenecks exist in a typical
+calculation; by focusing on optimized C++ implementations of these bottlenecks
+and making these C++ functions available in Python, most of the code to implement
+the overall calculation can be written in simpler Python code.  Python is far
+better suited to management tasks such as directory navigation and retrieval,
+making it a natural choice for overall calculation layout than C++.  With the
+emergence of :numpy:`NumPy <>` as a standard tool for executing almost any
```
Comment (robertodr) [inline doc/sphinxman/source/prog_ways_to_add.rst]: I think there's a missing link to NumPy here.

File: doc/sphinxman/source/prog_ways_to_add.rst [hunk 8]
```diff
@@ -35,15 +35,106 @@ Ways to Add Code: Psi4NumPy, Plugins, Full Integration
...
+C++ to Python has facilitated a much simpler work flow for prototyping and
+developing methods: this is detailed in the next section.
 
-There is a separate section that more fully explains plugins.
+.. _`sec:prog_psi4numpy`:
 
+Rapid initial development using Psi4NumPy
+-----------------------------------------
 
+The :p4numpy:`Psi4NumPy <>` project [Smith:2018:3504]_ is the recommended
```
Comment (robertodr) [inline doc/sphinxman/source/prog_ways_to_add.rst]: Missing link to Psi4NumPy repo

File: doc/sphinxman/CMakeLists.txt (hunk 6 above)
Comment (loriab) [reply doc/sphinxman/CMakeLists.txt]: `tests/CMakeLists.txt` is the choicer target

File: doc/sphinxman/source/prog_tour.rst [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+its members available to the Python layer, any other similar export functions
+should be written in python.
+
+Integral consuming technologies
+-------------------------------
+
+The general philosophy in |PSIfour| is to try write two-electron integral
+driven tasks in methods like Hartree-Fock, CIS and CPHF in terms of generalized
+Fock-like matrices.  From here, a single class can be used to construct these
+generalized Fock matrices, which is what libFock
```
[the remark this answers is not in the document]
Comment (JonathonMisiewicz) [reply doc/sphinxman/source/prog_tour.rst]: Seconding this. I've only ever heard of a generalized Fock matrix in terms of orbital optimization. Looking over the Yamaguchi text, you're talking about what I recognize as reduced density matrices, maybe? But then this wouldn't make any sense for theories like CC where we need to do computations before we have the lambda's to make the RDMs. Do you mean general quantities that one could "einsum" over in Python?

**Comment by JonathonMisiewicz** [comment]
Some "general direction" comments:

1. I can't identify what the "Ways to Add Code" page is supposed to accomplish. There's nothing unifying these sections together, and even individual sections seem disjointed to me. The easier and more rapid development section feels like one part expositing historical developments and one part trying to sell the new developer on the transition to Python. The Psi4Numpy and plugins sections read like one part overviewing the process of going from equations to code and again like one part trying to sell the new developer on NumPy. The "incorporating code into Psi4" section reads like a guide on parceling code into C++ parts and Python parts.

2. Are you expecting the reader to already know about dual-language software and OpenMP? The current documentation is written as if you are, but I didn't know about the first one until after I had started working in Psi, and I still only have a hazy idea on the second.

3. I fear that the strategy of "mostly the vanilla contributor covenant, but with small additions to emphasize that we value inexperienced devs" will lead to those small additions not being noticed by the inexperienced devs. When I see something as general as, "to help us maintain this vibrant
and productive atmosphere," my immediate thought is that this is administrative babble and only worth skimming. Having read the covenant, I cannot identify the small additions in there. Might I suggest making an explicit section on new developers and making it very specific to Psi? Something like:
_"The continued maintenance and growth of Psi depends on getting, training, and retaining new developers. This brings in prospective developers with widely varying levels of programming experience, including people learning the basics of theory and programming in quantum chemistry, people learning the basics of C++ or Python, and even people learning programming in general. We accept all this and are committed to training new developers to contribute to the code. In particular, we are willing to..."_

**Comment by dgasmith** [comment]
@andysim Any blockers to merging this in sooner rather than later? It seems quite a large contribution and is fairly stable, issues can be made and cleaned up over time.

**Comment by andysim** [comment]
Sorry, I ignored this for a while.  I agree with getting it in now and iterating on it over time.  I do want to address the comments/corrections first though.  I can get that done by the end of the day

**Comment by dgasmith** [comment]
Going to poke here again, something going in is better than the current state.

**Comment by dgasmith** [comment]
@andysim Gentle reminder here, I think what is currently extant is very useful even if not yet perfect.

**Pull request closed without merging**